### PR TITLE
[IMP] project: add button to hide closed sub-tasks

### PR DESCRIPTION
**Pull request by bastvdn**
project.task form view > sub-tasks notebook > add a 'hide closed tasks' button (in the same style as 'add a line')

Small remark for the JS tests =>
The behavior of this widget is related to the currentStorage (simply remembers the previous configuration)
i do not reset them in the tests so if you run the same Qunit test again in your browser, it'll fails.
It seems to work on the runbot but it will be better with localStorage reset before the test.
I tried those with a few ways but couldn't succeed to reset the localStorage consistently.

Task-3703563

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.xml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+    <t t-name="project.NotebookTaskListRenderer.Rows" t-inherit="web.ListRenderer.Rows" t-inherit-mode="primary">
+        <xpath expr="//t[@t-foreach='list.records']" position="replace">
+            <t t-foreach="list.records" t-as="record" t-key="record.id">
+                <t t-if="!['1_done','1_canceled'].includes(record.data.state) or !hideState.hide">
+                    <t t-call="{{ constructor.recordRowTemplate }}"/>
+                </t>
+            </t>
+        </xpath>
```
Comment (pkyriakou) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.xml:10]: Just visual stuff :nerd_face: 
```suggestion
        </xpath>
        
```

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    get _getStorageKey() {
+        return `hide_closed_${this.constructor.name}`;
+    }
+
+    get hideClosed() {
+        return this.hideState.hide
+    }
+
+    get ShowX2MRecords() {
+        return this.props.list.context.closed_X2M_count > 0 || typeof this.props.list.context.closed_X2M_count === 'undefined';
```
Comment (pkyriakou) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:29]: why are we showing the buttons when the second condition is met again ? kinda confused with this :thinking:

File: addons/project/views/project_task_views.xml [hunk 3]
```diff
@@ -438,7 +440,7 @@
                                     <field name="company_id" column_invisible="True"/>
                                     <field name="date_deadline" invisible="state in ['1_done', '1_canceled']" optional="hide" decoration-danger="date_deadline &lt; current_date"/>
                                     <field name="activity_ids" string="Next Activity" widget="list_activity" optional="hide"/>
-                                     <field name="my_activity_date_deadline" string="My Deadline" widget="remaining_days" options="{'allow_order': '1'}" optional="hide"/>
+                                    <field name="my_activity_date_deadline" string="My Deadline" widget="remaining_days" options="{'allow_order': '1'}" optional="hide"/>
```
Comment (pkyriakou) [inline addons/project/views/project_task_views.xml:443]: unnecessary diff ? :thinking:

File: addons/project/static/tests/project_notebook_task_list_tests.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        ]);
+        pyEnv['project.task'].create([
+            { name: 'task one', project_id: projectId, closed_subtask_count: 1, subtask_count: 4, child_ids: [2, 3, 4, 7], depend_on_ids: [5,6], state: '04_waiting_normal', user_ids: [userId] },
+            { name: 'task two', parent_id: 1, closed_subtask_count: 0, subtask_count: 0, child_ids: [], depend_on_ids: [], state: '03_approved' },
+            { name: 'task three', parent_id: 1, closed_subtask_count: 0, subtask_count: 0, child_ids: [], depend_on_ids: [], state: '02_changes_requested' },
+            { name: 'task four', parent_id: 1, closed_subtask_count: 0, subtask_count: 0, child_ids: [], depend_on_ids: [], state: '1_done' },
+            { name: 'task five', closed_subtask_count: 0, subtask_count: 1, child_ids: [6], depend_on_ids: [], state: '03_approved' },
+            { name: 'task six', parent_id: 5, closed_subtask_count: 0, subtask_count: 0, child_ids: [], depend_on_ids: [], state: '1_canceled' },
+            { name: 'task seven', parent_id: 1, closed_subtask_count: 0, subtask_count: 0, child_ids: [], depend_on_ids: [], state: '01_in_progress', user_ids: [userId] },
```
Comment (pkyriakou) [inline addons/project/static/tests/project_notebook_task_list_tests.js:34]: do you need to specify the closed_subtask_count / subtask_count ? are they not computed fields ?

File: addons/project/static/tests/project_notebook_task_list_tests.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            res_model: "project.task",
+            views: [[false, "form"]],
+            res_id: 1,
+        });
+
+        assert.containsN(target, 'div[name="child_ids"] .o_data_row', 4, "The subtasks list should display all subtasks by default, thus we are looking for 4 in total");
+        assert.containsN(target, 'div[name="depend_on_ids"] .o_data_row', 2, "The depend on tasks list should display all blocking tasks by default, thus we are looking for 2 in total");
+
+        await contains("div[name='child_ids'] .o_field_x2many_list_row_add a.o_toggle_closed_task_button", { text: "Hide closed tasks" });
+        await contains("div[name='depend_on_ids'] .o_field_x2many_list_row_add a.o_toggle_closed_task_button", { text: "Hide closed tasks" });
```
Comment (pkyriakou) [inline addons/project/static/tests/project_notebook_task_list_tests.js:75]: you could use regular assertions here instead of contains :thinking: if we are doing this for checking that the correct label is printed I think assertions are more sound - the same applied below.

**Review by pkyriakou — COMMENTED** [review]
Thank you very much for your work @bastvdn :100: 
I left some comments for you to look into :v:

File: addons/project/views/project_task_views.xml (hunk 3 above)
Comment (bastvdn) [reply addons/project/views/project_task_views.xml:443]: meh, just fixing and indent :shrug: i can remove this if you want

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js (hunk 2 above)
Comment (bastvdn) [reply addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:29]: Mmh yeah i guessed that it should always be possible to display the button even if the current x2m task field doesn't have any "closed_X2M_count" fields that could track the number of closed task in this x2m dynamically.

For exemple in this PR, i needed to add a closed_depend_on_count computed field for the blocked by task field count, i don't think this should be necessary for every potential x2m task field displayed in a tree view

So if i don't specify the closed_X2M_count we always display the Show/Hide button
Tell me if that makes sense for you

File: addons/project/models/project_task.py [hunk 6]
```diff
@@ -475,16 +476,23 @@ def _compute_recurring_count(self):
     def _compute_depend_on_count(self):
         tasks_with_dependency = self.filtered('allow_task_dependencies')
         (self - tasks_with_dependency).depend_on_count = 0
+        (self - tasks_with_dependency).closed_depend_on_count = 0
+        if not any(self._ids):
+            for task in self:
+                task.depend_on_count, task.closed_depend_on_count = len(task.depend_on_ids), len(task.depend_on_ids.filtered(lambda r: r.state in CLOSED_STATES))
```
Comment (pkyriakou) [inline addons/project/models/project_task.py:482]: honestly I'd split this into two lines :shrug: I see no reason why it should be one line :thinking: 
```suggestion
                task.depend_on_count = len(task.depend_on_ids)
                task.closed_depend_on_count = len(task.depend_on_ids.filtered(lambda r: r.state in CLOSED_STATES))
```

**Review by pkyriakou — COMMENTED** [review]
Thank you very much @bastvdn for the updates and the answers :100: 
I left one very minor comment but since it's small I will green the task regardless. Also as we discussed, it would perhaps be nice to add a comment in the code to explain why we check whether the type of the ``closed_x2m_count`` is undefined.

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js (hunk 2 above)
Comment (bastvdn) [reply addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:31]: TLDR: basically if the user does not set up this prop in the view we always display the "show closed tasks" button
Alternativelly we could do that count in the frontend and not use a counter field at all, but since it's already there for the subtasks i thought it would be ok to create another computed field.
image

File: addons/project/models/project_task.py (hunk 6 above)
Comment (bastvdn) [reply addons/project/models/project_task.py]: Thanks for the review :) added the comment

File: addons/project/models/project_task.py [hunk 7]
```diff
@@ -477,16 +478,24 @@ def _compute_recurring_count(self):
     def _compute_depend_on_count(self):
         tasks_with_dependency = self.filtered('allow_task_dependencies')
         (self - tasks_with_dependency).depend_on_count = 0
+        (self - tasks_with_dependency).closed_depend_on_count = 0
+        if not any(self._ids):
```
Comment (MissingNoShiny) [inline addons/project/models/project_task.py:482]: Shouldn't it be if any record is new, rather than if no record is not new? I guess in practice, it won't make much of a difference, I don't think the compute could be called on a mix of both?

File: addons/project/static/tests/project_notebook_task_list_tests.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
+/* @odoo-module */
```
Comment (MissingNoShiny) [inline addons/project/static/tests/project_notebook_task_list_tests.js:1]: I think we should probably convert the tests to Hoot 😬

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import { ListRenderer } from '@web/views/list/list_renderer';
+
+import { useState } from "@odoo/owl";
+
+export class NotebookTaskListRenderer extends ListRenderer {
+    static rowsTemplate = "project.NotebookTaskListRenderer.Rows";
+    setup() {
+        super.setup();
+        this.hideState = useState({
+            hide: localStorage.getItem(this._getStorageKey) == 'true' || false,
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:12]:
```suggestion
            hide: localStorage.getItem(this._getStorageKey) === 'true',
```

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    get hideClosed() {
+        return this.hideState.hide
+    }
+
+    get closedX2MCount() {
+        return this.props.list.context.closed_X2M_count;
+    }
+
+    get openLabel() {
+        return `${typeof this.closedX2MCount === 'undefined' ? 'Show' : this.closedX2MCount} closed task` + (this.closedX2MCount > 1 ? "s" : "");
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:33]: We should make this translatable

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    get closeLabel() {
+        return 'Hide closed task' + (this.closedX2MCount > 1 ? "s" : "");
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:37]: Same here

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    get toggleLabel() {
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:40]: We could have a more explicit name for this getter and the two above

**Review by MissingNoShiny — CHANGES_REQUESTED** [review]
Thanks, I left a few comments 🙂

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    /**
+     * @private
+     * @returns {string}
+     */
+    get _getStorageKey() {
+        return `hide_closed_${this.constructor.name}`;
+    }
+
+    get hideClosed() {
+        return this.hideState.hide
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:27]:
```suggestion
        return this.hideState.hide;
```

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    get closedX2MCount() {
+        return this.props.list.context.closed_X2M_count;
+    }
+
+    get openLabel() {
+        return this.closedX2MCount === undefined ? _t("Show closed tasks") : _t("%s closed tasks", this.closedX2MCount);
+    }
+
+    get closeLabel() {
+        return _t('Hide closed tasks');
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:39]:
```suggestion
        return _t("Hide closed tasks");
```

File: addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    get toggleListHideLabel() {
+        return this.hideClosed ? this.openLabel : this.closeLabel;
+    }
+
+    get ShowX2MRecords() {
+        // If there isn't a closed_X2M_count defined in the context of the x2m task in the view we are always displaying the Toggle button
+        // In case there is no computed field to calculate the number of closed X2M tasks in the backend
+        return this.closedX2MCount > 0 || typeof this.closedX2MCount === 'undefined';
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/notebook_task_one2many_field/notebook_task_list_renderer.js:49]:
```suggestion
        return this.closedX2MCount > 0 || typeof this.closedX2MCount === "undefined";
```

File: addons/project/static/src/components/subtask_one2many_field/subtask_list_renderer.js [hunk 16]
```diff
@@ -2,9 +2,10 @@
 
 import { _t } from "@web/core/l10n/translation";
 import { ConfirmationDialog } from "@web/core/confirmation_dialog/confirmation_dialog";
-import { TaskListRenderer } from "../task_list_renderer";
+import { NotebookTaskListRenderer } from '../notebook_task_one2many_field/notebook_task_list_renderer';
 
-export class SubtaskListRenderer extends TaskListRenderer {
+
```
Comment (MissingNoShiny) [inline addons/project/static/src/components/subtask_one2many_field/subtask_list_renderer.js:7]:
```suggestion
```

**Review by MissingNoShiny — COMMENTED** [review]
Just a few linting comments :slightly_smiling_face:

**Review by MissingNoShiny — APPROVED** [review]
LGTM, thanks! 🙂 
robodoo r+

**Comment by MissingNoShiny** [comment]
robodoo r+
